File: sentinel/src/telemetry.ts
/**
 * Guardian telemetry sender.
 *
 * Posts `IncidentEvent`s to Guardian's `/api/ingest` endpoint per the wire
 * format in `docs/specs/wire-format.md`.
 *
 * Best-effort by design:
 *   - Never blocks the user's model call. Fire-and-forget.
 *   - Catches every error. Failures are silent at v0.1 — a future option
 *     may surface them via `onIncident` with a `[telemetry-error]` event.
 *   - Bounded retries with exponential backoff for transient failures
 *     (5xx, network errors). 4xx (validation/auth) is permanent — drop.
 *   - Bounded queue. We never accumulate more than `MAX_QUEUE` pending
 *     sends — past that, we drop newest and log a warning. Keeps memory
 *     stable under outage.
 *
 * v0.1: one event per envelope (no batching). Batching ships in v0.2.
 */

import type { GuardianConfig, IncidentEvent } from "./types.js";
import { AGENT_ID } from "./agent-id.js";
import { SENTINEL_VERSION } from "./version.js";

const ENVELOPE_VERSION = "1" as const;
const LICTOR_CORE_VERSION = "0.1.0";
const DEFAULT_TIMEOUT_MS = 5000;
const MAX_RETRIES = 3;
const BASE_BACKOFF_MS = 200;

/** Cap on in-flight + pending sends. Past this, newest sends are dropped. */
const MAX_QUEUE = 100;

interface Envelope {
  envelope_version: typeof ENVELOPE_VERSION;
  sentinel_version: string;
  lictor_core_version: string;
  agent_id: string;
  sent_at: string;
  events: IncidentEvent[];
}

/** Build an envelope around a single event. Pure — no I/O. */
export function buildEnvelope(event: IncidentEvent): Envelope {
  return {
    envelope_version: ENVELOPE_VERSION,
    sentinel_version: SENTINEL_VERSION,
    lictor_core_version: LICTOR_CORE_VERSION,
    agent_id: AGENT_ID,
    sent_at: new Date().toISOString(),
    events: [event],
  };
}

/** Pluggable fetch — overridden in tests to assert request shape without real network. */
export interface FetchLike {
  (input: string, init: RequestInit): Promise<{ ok: boolean; status: number; text(): Promise<string> }>;
}

let _fetch: FetchLike = globalThis.fetch as unknown as FetchLike;

/** Override the fetch used by sendToGuardian. Tests call this. */
export function setFetchForTests(f: FetchLike | null): void {
  _fetch = (f ?? (globalThis.fetch as unknown as FetchLike));
}

let _inFlight = 0;

/** Number of pending sends, for tests. */
export function pendingCount(): number {
  return _inFlight;
}

/**
 * Send a single event to Guardian. Best-effort, fire-and-forget.
 *
 * Retries:
 *   - 5xx / network error → retry up to MAX_RETRIES with exponential backoff
 *     (200ms, 400ms, 800ms — jittered ±20%)
 *   - 4xx → permanent failure, no retry (validation / auth)
 *   - 2xx → success
 */
export async function sendToGuardian(
  event: IncidentEvent,
  config: GuardianConfig,
): Promise<void> {
  if (_inFlight >= MAX_QUEUE) {
    // Drop. Don't accumulate unbounded under outage.
    return;
  }
  _inFlight++;
  try {
    await sendWithRetry(event, config);
  } finally {
    _inFlight--;
  }
}

async function sendWithRetry(event: IncidentEvent, config: GuardianConfig): Promise<void> {
  const envelope = buildEnvelope(event);
  const body = JSON.stringify(envelope);
  const timeoutMs = config.timeoutMs ?? DEFAULT_TIMEOUT_MS;

  for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), timeoutMs);

    try {
      const r = await _fetch(config.endpoint, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "Authorization": `Bearer ${config.token}`,
        },
        body,
        signal: ctrl.signal,
      });
      clearTimeout(timer);

      if (r.ok) return; // 2xx — done.

      // 4xx: permanent. No retry.
      if (r.status >= 400 && r.status < 500) {
        return;
      }
      // 5xx: fall through to retry.
    } catch {
      clearTimeout(timer);
      // Network error / abort — fall through to retry.
    }

    if (attempt < MAX_RETRIES) {
      const backoff = BASE_BACKOFF_MS * Math.pow(2, attempt);
      const jitter = backoff * (0.8 + Math.random() * 0.4);
      await new Promise((r) => setTimeout(r, jitter));
    }
  }
  // All retries exhausted. Drop silently.
}

File: sentinel-py/src/lictor_sentinel/telemetry.py
"""
Guardian telemetry sender (Python).

Mirrors @lictor/sentinel/src/telemetry.ts. Same wire format, same retry
policy, same privacy invariants (never ships raw user content). Uses
urllib (stdlib) — zero runtime dependencies.

Best-effort by design: failures never block the user's model call.
"""

from __future__ import annotations

import dataclasses
import json
import random
import time
import urllib.error
import urllib.request
from typing import Callable

from .agent_id import AGENT_ID
from .types import GuardianConfig, IncidentEvent

ENVELOPE_VERSION = "1"
LICTOR_CORE_VERSION = "0.1.0"
SENTINEL_VERSION = "0.1.0a0"

_DEFAULT_TIMEOUT_S = 5.0
_MAX_RETRIES = 3
_BASE_BACKOFF_S = 0.2
_MAX_QUEUE = 100


def _event_to_wire(e: IncidentEvent) -> dict:
    """Serialize an IncidentEvent dataclass to the wire-format JSON shape.

    Note: wire format uses camelCase (matches TS); Python dataclass uses
    snake_case (Python convention). This is the translation layer.
    """
    return {
        "ts": e.ts,
        "agentId": e.agent_id,
        "phase": e.phase,
        "checkId": e.check_id,
        "severity": e.severity,
        "title": e.title,
        "detail": e.detail,
        "model": dataclasses.asdict(e.model),
        "fingerprint": e.fingerprint,
        "action": e.action,
        "sentinelVersion": e.sentinel_version,
    }


def build_envelope(event: IncidentEvent) -> dict:
    """Build a wire-format envelope around a single event."""
    return {
        "envelope_version": ENVELOPE_VERSION,
        "sentinel_version": SENTINEL_VERSION,
        "lictor_core_version": LICTOR_CORE_VERSION,
        "agent_id": AGENT_ID,
        "sent_at": _now_iso(),
        "events": [_event_to_wire(event)],
    }


def _now_iso() -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%S.000Z", time.gmtime())


# Pluggable transport for tests. Returns (status, body_text).
# Default uses urllib; tests inject a mock.
Transport = Callable[[str, bytes, dict[str, str], float], tuple[int, str]]


def _urllib_transport(url: str, body: bytes, headers: dict[str, str], timeout: float) -> tuple[int, str]:
    req = urllib.request.Request(url, data=body, headers=headers, method="POST")
    try:
        with urllib.request.urlopen(req, timeout=timeout) as r:
            return r.status, r.read(2 * 1024 * 1024).decode("utf-8", errors="replace")
    except urllib.error.HTTPError as e:
        return e.code, ""


_transport: Transport = _urllib_transport
_in_flight = 0


def set_transport_for_tests(t: Transport | None) -> None:
    """Override the transport. Tests use this to inject a mock."""
    global _transport
    _transport = t if t is not None else _urllib_transport


def pending_count() -> int:
    """Number of in-flight sends (for tests / health checks)."""
    return _in_flight


def send_to_guardian(event: IncidentEvent, config: GuardianConfig) -> None:
    """
    Send a single event to Guardian. Best-effort, synchronous (Python's
    sync model — the application can run this in a thread / background
    task if it wants to avoid blocking the model call).

    Retry policy mirrors the TS SDK:
      - 5xx / network error → up to 3 retries with exponential backoff
      - 4xx → permanent failure, no retry
      - 2xx → success
    """
    global _in_flight
    if _in_flight >= _MAX_QUEUE:
        return  # drop; never accumulate unbounded under outage
    _in_flight += 1
    try:
        _send_with_retry(event, config)
    finally:
        _in_flight -= 1


def _send_with_retry(event: IncidentEvent, config: GuardianConfig) -> None:
    body_dict = build_envelope(event)
    body = json.dumps(body_dict).encode("utf-8")
    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {config.get('token', '')}",
    }
    timeout = config.get("timeout_ms", _DEFAULT_TIMEOUT_S * 1000) / 1000

    for attempt in range(_MAX_RETRIES + 1):
        try:
            status, _ = _transport(
                config.get("endpoint", ""),
                body,
                headers,
                timeout,
            )
            if 200 <= status < 300:
                return  # success
            if 400 <= status < 500:
                return  # permanent
            # 5xx: fall through to retry
        except Exception:
            # Network error: fall through to retry
            pass

        if attempt < _MAX_RETRIES:
            backoff = _BASE_BACKOFF_S * (2**attempt)
            jitter = backoff * (0.8 + random.random() * 0.4)
            time.sleep(jitter)
    # All retries exhausted.
